source/common/grpc/transcoding_filter.cc

```diff
@@ -0,0 +1,396 @@
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> request_stream_;
+  std::unique_ptr<::google::grpc::transcoding::TranscoderInputStream> response_stream_;
+};
+
+} // namespace
+
+TranscodingConfig::TranscodingConfig(const Json::Object& config) {
+  std::string proto_descriptor_file = config.getString("proto_descriptor");
+  FileDescriptorSet descriptor_set;
+  if (!descriptor_set.ParseFromString(Filesystem::fileReadToEnd(proto_descriptor_file))) {
+    throw EnvoyException("Unable to parse proto descriptor");
+  }
+
+  for (const auto& file : descriptor_set.file()) {
+    if (descriptor_pool_.BuildFile(file) == nullptr) {
+      throw EnvoyException("Unable to parse proto descriptor");
+    }
+  }
+
+  google::grpc::transcoding::PathMatcherBuilder<const google::protobuf::MethodDescriptor*> pmb;
+
+  for (const auto& service_name : config.getStringArray("services")) {
+    auto service = descriptor_pool_.FindServiceByName(service_name);
+    if (service == nullptr) {
+      throw EnvoyException("Could not find '" + service_name + "' in the proto descriptor");
+    }
+    for (int i = 0; i < service->method_count(); ++i) {
+      auto method = service->method(i);
+
+      auto http_rule = method->options().GetExtension(google::api::http);
+
+      log().debug("/" + service->full_name() + "/" + method->name());
+      log().debug(http_rule.DebugString());
+
+      switch (http_rule.pattern_case()) {
+      case ::google::api::HttpRule::kGet:
+        pmb.Register("GET", http_rule.get(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPut:
+        pmb.Register("PUT", http_rule.put(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPost:
+        pmb.Register("POST", http_rule.post(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kDelete:
+        pmb.Register("DELETE", http_rule.delete_(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPatch:
+        pmb.Register("PATCH", http_rule.patch(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kCustom:
+        pmb.Register(http_rule.custom().kind(), http_rule.custom().path(), http_rule.body(),
+                     method);
+        break;
+      default:
+        break;
+      }
+    }
+  }
+
+  path_matcher_ = pmb.Build();
+
+  type_helper_.reset(new google::grpc::transcoding::TypeHelper(
+      google::protobuf::util::NewTypeResolverForDescriptorPool(kTypeUrlPrefix, &descriptor_pool_)));
+
+  log().debug("transcoding filter loaded");
+}
+
+Status TranscodingConfig::CreateTranscoder(
+    const Http::HeaderMap& headers, ZeroCopyInputStream* request_input,
+    ::google::grpc::transcoding::TranscoderInputStream* response_input,
+    std::unique_ptr<Transcoder>& transcoder,
+    const google::protobuf::MethodDescriptor*& method_descriptor) {
+  std::string method = headers.Method()->value().c_str();
```

Comment:
Please make all these temporary variables const unless they are going to be mutated.